scheduler/feasible.go

```diff
@@ -177,6 +185,75 @@ func (h *HostVolumeChecker) hasVolumes(n *structs.Node) bool {
 	return true
 }
 
+type CSIVolumeChecker struct {
+	ctx     Context
+	volumes map[string]*structs.VolumeRequest
+}
+
+func NewCSIVolumeChecker(ctx Context) *CSIVolumeChecker {
+	return &CSIVolumeChecker{
+		ctx: ctx,
+	}
+}
+
+func (c *CSIVolumeChecker) SetVolumes(volumes map[string]*structs.VolumeRequest) {
+	c.volumes = volumes
+}
+
+func (c *CSIVolumeChecker) Feasible(n *structs.Node) bool {
+	if c.hasPlugins(n) {
+		return true
+	}
+
+	c.ctx.Metrics().FilterNode(n, FilterConstraintCSIVolumes)
+	return false
+}
+
+func (c *CSIVolumeChecker) hasPlugins(n *structs.Node) bool {
+	// Have volume if
+	// - healthy controller plugin is running the driver
+	// - the volume has free claims
+	// - the node is running the node plugin, implies matching topology
+
+	rLen := len(c.volumes)
+	hLen := len(n.CSINodePlugins)
+
+	// Fast path: Requested no volumes. No need to check further.
+	if rLen == 0 {
+		return true
+	}
+
+	// Fast path: Requesting more volumes than the node has, can't meet the criteria.
+	if rLen > hLen {
```

Comment:
c.volumes here is a map of volume request name to volume request info, so the later comparison against lengths doesn't really make sense to me?

What would make sense would be `if numberOfVolumesForPlugin + len(attachedVolumesOnHostForPlugin) > nodePluginMaxVolumes`?

Reply:
ugh, sorry, I bootstrapped this by copying from the host volume checker, and it's just stale. That check you just suggested makes sense, but only per plugin, so it's not really a fast path anymore.

Reply:
@endocrimes there's a little bit to this, so I created https://github.com/hashicorp/nomad/issues/7025 to keep track of it. If it's ok, I'd like to merge this so we can start using it.

Reply:
Sgtm 👍